consensus/core/src/linearizer.rs

```diff
@@ -0,0 +1,491 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::HashSet,
+    fmt::{self, Display, Formatter},
+    sync::Arc,
+};
+
+use parking_lot::RwLock;
+
+use crate::{
+    block::{BlockAPI, BlockRef, BlockTimestampMs, VerifiedBlock},
+    commit::{Commit, CommitIndex},
+    dag_state::DagState,
+    storage::Store,
+};
+
+/// The output of consensus is an ordered list of [`CommittedSubDag`]. The application
+/// can arbitrarily sort the blocks within each sub-dag (but using a deterministic algorithm).
+#[derive(Clone)]
+pub struct CommittedSubDag {
+    /// A reference to the leader of the sub-dag
+    pub leader: BlockRef,
+    /// All the committed blocks that are part of this sub-dag
+    pub blocks: Vec<VerifiedBlock>,
+    /// The timestamp of the commit, obtained from the timestamp of the leader block.
+    pub timestamp_ms: BlockTimestampMs,
+    /// Index of the commit.
+    /// First commit after genesis has a index of 1, then every next commit has a
+    /// index incremented by 1.
+    pub commit_index: CommitIndex,
+}
+
+#[allow(unused)]
+impl CommittedSubDag {
+    /// Create new (empty) sub-dag.
+    pub fn new(
+        leader: BlockRef,
+        blocks: Vec<VerifiedBlock>,
+        timestamp_ms: u64,
+        commit_index: CommitIndex,
+    ) -> Self {
+        Self {
+            leader,
+            blocks,
+            timestamp_ms,
+            commit_index,
+        }
+    }
+
+    pub fn new_from_commit_data(commit_data: Commit, block_store: Arc<dyn Store>) -> Self {
```

Comment:
In general `DagState` should be able to replace block store at each callsite, except when we need to look up arbitrary committed block. This function should work on `DagState` too.